--- /usr/ia16-elf/lib/dos-ms.ld	2024-02-18 23:02:32.000000000 +0100
+++ bin/andrea-module.ld	2024-03-21 17:29:23.200227027 +0100
@@ -6,6 +6,7 @@
 
 OUTPUT_FORMAT(binary)
 ENTRY(_start)
+INPUT(-l:andrea-modstart.a)
 
 MEMORY
   {
@@ -241,7 +242,7 @@
 		/* Header size in paragraphs.  */
 		SHORT (__msdos_mz_hdr_paras)
 		/* Minimum extra paragraphs.  */
-		SHORT ((0x10000 - SIZEOF (.data) - ADDR (.data)) / 16)
+		SHORT ((__lbss0 + 15) / 16 + 1)
 		/* Maximum extra paragraphs.
 
 		   In the normal case, instead of setting this to 0xffff so
@@ -256,12 +257,11 @@
 		   try to place our .exe at the top of conventional memory,
 		   which will crash everything).  -- tkchia  */
 		SHORT (DEFINED (__msdos_handle_v1) ? 0xffff
-		    : (0x10000 - SIZEOF (.data) - ADDR (.data)) / 16)
-		/* Initial %ss.  */
-		SHORT (LOADADDR (.data) / 16 - __msdos_mz_hdr_paras
-		    - ADDR (.data) / 16 + 0x10000)
-		/* Initial %sp.  Let it wrap around from 0.  */
-		SHORT (0)
+		    : ((__lbss0 + 15) / 16 + 1))
+		/* Initial %ss, just after the .bss section.  */
+		SHORT (LOADADDR (.bss) / 16 - ADDR (.bss) / 16)
+		/* Initial %sp. One paragraph is more than enough for an error message.  */
+		SHORT (0x10)
 		/* Checksum (unused).  */
 		SHORT (0)
 		/* Initial %cs:%ip.  */
@@ -297,6 +297,9 @@
     /* Target text sections.  */
     .text : {
 		__stext = .;
+
+		INCLUDE ../include/andrea/header.ld
+
 		*(.startupA)
 		*(.startupB)
 		*(.startupC)
@@ -312,6 +315,9 @@
 		*(.fini)
 		*(.exitB)
 		*(.text) *(.text.*)
+
+		/* Module data (tiny-ish memory model).  */
+		*(.andrea.moddata)
 		__etext = .;
 
 		/* Make the data segments start on a paragraph boundary.  */
